xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -0,0 +1,299 @@
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Serves as a wrapper for {@link XdsClientImpl} used on the server side by {@link
+ * io.grpc.xds.internal.sds.XdsServerBuilder}.
+ */
+@Internal
+public final class XdsClientWrapperForServerSds {
+  private static final Logger logger =
+      Logger.getLogger(XdsClientWrapperForServerSds.class.getName());
+
+  private static final EventLoopGroupResource eventLoopGroupResource =
+      Epoll.isAvailable() ? new EventLoopGroupResource("GrpcServerXdsClient") : null;
+  private final InternalLogId logId;
+  // Must be accessed from the syncContext
+  private boolean panicMode;
+  final SynchronizationContext syncContext =
+      new SynchronizationContext(
+        new Thread.UncaughtExceptionHandler() {
+          @Override
+          public void uncaughtException(Thread t, Throwable e) {
+            logger.log(
+                Level.SEVERE,
+                  "[" + getLogId() + "] Uncaught exception in the SynchronizationContext. Panic!",
+                  e);
+            panic(e);
+          }
+        });
+
+  private EnvoyServerProtoData.Listener curListener;
+  @Nullable private ObjectPool<XdsClient> xdsClientPool;
+  @Nullable private XdsClient xdsClient;
+
+  /** Wraps the XdsClientImpl for use by SdsProtocolNegotiators. */
+  public XdsClientWrapperForServerSds(int port, Bootstrapper bootstrapper) {
+    this.logId = InternalLogId.allocate("GrpcServer", Integer.toString(port));
+    Bootstrapper.BootstrapInfo bootstrapInfo = null;
+    try {
+      bootstrapInfo = bootstrapper.readBootstrap();
+    } catch (IOException e) {
+      logger.log(Level.SEVERE, "Error from readBootstrap", e);
+      return;
+    }
+    final List<Bootstrapper.ServerInfo> serverList = bootstrapInfo.getServers();
+    final Node node = bootstrapInfo.getNode();
+    if (serverList.isEmpty()) {
+      logger.log(Level.SEVERE, "No traffic director provided by bootstrap");
+      return;
+    }
+    final EventLoopGroup timeService = SharedResourceHolder.get(eventLoopGroupResource);
+
+    XdsClient.XdsClientFactory xdsClientFactory =
+        new XdsClient.XdsClientFactory() {
+          @Override
+          XdsClient createXdsClient() {
+            return new XdsClientImpl(
+                "",
+                serverList,
+                XdsClient.XdsChannelFactory.getInstance(),
+                node,
+                syncContext,
+                timeService,
+                new ExponentialBackoffPolicy.Provider(),
+                GrpcUtil.STOPWATCH_SUPPLIER);
+          }
+        };
+    xdsClientPool = new XdsClient.RefCountedXdsClientObjectPool(xdsClientFactory);
```

Comment:
Why you use an object pool? You are just creating an `XdsClient` instance in-place without sharing with other classes.

Reply:
You are right and my original prototype implementation didn't have the pool. When I adapted to the new XdsClientImpl I ended up using the existing pattern. I don't think there will be other classes needing to use a "shared" XdsClient on the server side so I will change this to the non-pool implementation.